File: packages/renderer-ios/ios/Sources/TerajsRendererHost/TerajsHostNode.swift
import UIKit

public protocol UIKitHostNode: AnyObject {
  var nodeId: Int { get }
  var parentId: Int? { get set }
}

public final class UIKitHostElementNode: UIKitHostNode {
  public let nodeId: Int
  public let viewType: String
  public let view: UIView
  public var parentId: Int?
  public var childNodeIds: [Int]
  public var props: [String: TerajsJSONValue]
  public var styles: [String: String]
  public var className: String?
  public var subscribedEvents: Set<String>

  init(nodeId: Int, viewType: String, view: UIView) {
    self.nodeId = nodeId
    self.viewType = viewType
    self.view = view
    self.childNodeIds = []
    self.props = [:]
    self.styles = [:]
    self.subscribedEvents = []
  }
}

public final class UIKitHostTextNode: UIKitHostNode {
  public let nodeId: Int
  public var parentId: Int?
  public var value: String

  init(nodeId: Int, value: String) {
    self.nodeId = nodeId
    self.value = value
  }
}

File: packages/renderer-android/android/src/main/kotlin/dev/terajs/renderer/android/TerajsHostNode.kt
package dev.terajs.renderer.android

import android.view.View
import android.widget.TextView

sealed interface AndroidHostNode {
  val nodeId: Int
  var parentId: Int?
}

data class AndroidHostElementNode(
  override val nodeId: Int,
  val viewType: String,
  val view: View,
  override var parentId: Int? = null,
  val childNodeIds: MutableList<Int> = mutableListOf(),
  val props: MutableMap<String, TerajsJsonValue> = linkedMapOf(),
  val styles: MutableMap<String, String> = linkedMapOf(),
  var className: String? = null,
  val subscribedEvents: MutableSet<String> = linkedSetOf()
) : AndroidHostNode

data class AndroidHostTextNode(
  override val nodeId: Int,
  var value: String,
  val view: TextView,
  override var parentId: Int? = null
) : AndroidHostNode